### DISCUSSION TITLE
What are the steps to add a custom manager locally?

### DISCUSSION THREAD

**Discussion by HenrikPoulsen** [renovatebot/renovate#9780] [Request Help] [answered]
The docs are sadly incomplete and only mention where the manager and datasource files should go.
That is not enough for Renovate to pick up the changes.
Through spending a bunch of time trying to figure it out I found the `lib/manager/api.ts` file. Adding a reference to it there does start having it load.
Other discussions I have seen in the issues section seem to indicate that it should be enough to just drop them in the `lib/manager/some-custom-manager/` folder followed by building and it should be picked up.
"Is modifying api.ts a hard requirement to edit?" is my question basically.

My usecase is to make my own internal managers for custom datasources etc that are not public.
And I would like to create my own by modifying as little as possible inside Renovate, so that I can automate staying up to date with your newer releases without getting merge conflicts etc.

**Comment by rarkins** [comment, marked as answer]
You do need to add them to the `api.ts`. Having it built dynamically wasn't worth it for us so we returned to having a static import.

You're presumably aware we already have the regex managers so that you don't *need* to create custom managers for most cases. We also have plans for a custom datasource capability too, although I can't find an issue for it if we already have one. Feel free to contribute your ideas for making regex managers more manageable/scalable if you need, or for the custom datasources

**Reply by HenrikPoulsen** [reply]
Thanks for the reply. Yeah I do know about the regex managers and have been evaluating those.
The reason they are a tough sell is because we would be using this for thousands of repos on our enterprise github instance.
Propagating these regexes to all the repos and then keeping them updated does seem like quite the undertaking, since I haven't seen some mechanism for updating the renovate.json with new updated "default settings" after the onboard is complete.
So making our own managers and baking those into the images does seem like the path that will cause least overhead for us.
If you have some other ideas regarding that I am listening of course :)

**Reply by rarkins** [reply]
> Propagating these regexes to all the repos and then keeping them updated does seem like quite the undertaking, since I haven't seen some mechanism for updating the renovate.json with new updated "default settings" after the onboard is complete.

I'd recommend that you have an enterprise-wide preset that you use in the `onboardingConfig` so that all downstream repos can automatically get such "centralized" config changes immediately.

i.e.

- Create or pick a repo dedicated to presets (we use https://github.com/renovatebot/.github)
- Configure Renovate to extend *that* config preset by default when onboarding

**Reply by HenrikPoulsen** [reply]
Thanks that sounds promising. I'll look into that. Thanks for the help :)

**Comment by viceice** [comment]
You need a javascript file, which first adds new managers, datasources and then calls out main js entrypoint.

```js
const { getDatasources } = require ('renovate/dist/datasource');
const custom = require('./datasources/custom');

const ds = getDatasources();
ds.set(custom.id, custom.api);

require('renovate/dist/renovate');
```

But keep in mind this is internal api and isn't following our semver specs, so it can change even at a bugfix release. But we try our best to only do breakings on major versions.

**Reply by HenrikPoulsen** [reply]
That sounds pretty useful thanks. I'll play around with it for sure
